src/core/coap/coap.cpp

```diff
@@ -129,6 +134,165 @@ otError CoapBase::Send(ot::Message &aMessage, const Ip6::MessageInfo &aMessageIn
+
+    switch (aMessage.GetType())
+    {
+    case OT_COAP_TYPE_ACKNOWLEDGMENT:
+        // Check for block-wise transfer
+        if (aTransmitHook != NULL)
+        {
+            // Check for Block2 option in header
+            if (aMessage.ReadBlockOptionValues(OT_COAP_OPTION_BLOCK2) == OT_ERROR_NONE)
+            {
+                // Check if it is the first block of the transfer
+                if (aMessage.GetBlockWiseBlockNumber() == 0)
+                {
+                    // Set payload for first block of the transfer
+                    bufLen = otCoapBlockSizeFromExponent(aMessage.GetBlockWiseBlockSize());
+                    SuccessOrExit(error = aTransmitHook(aContext, buf, aMessage.GetBlockWiseBlockNumber() * bufLen,
+                                                        &bufLen, &moreBlocks));
+                    SuccessOrExit(error = aMessage.Append(buf, bufLen));
+
+                    CacheLastBlockResponse(&aMessage);
+                }
+            }
+        }
+
+        mResponsesQueue.EnqueueResponse(aMessage, aMessageInfo, aTxParameters);
+        break;
+    case OT_COAP_TYPE_RESET:
+        OT_ASSERT(aMessage.GetCode() == OT_COAP_CODE_EMPTY);
+        break;
+    default:
+        // Check for block-wise transfer
+        if (aTransmitHook != NULL)
+        {
+            // Check for Block2 option in header
+            if (aMessage.ReadBlockOptionValues(OT_COAP_OPTION_BLOCK1) == OT_ERROR_NONE)
+            {
+                // Check if it is the first block of the transfer
+                if (aMessage.GetBlockWiseBlockNumber() == 0)
+                {
+                    // Set payload for first block of the transfer
+                    bufLen = otCoapBlockSizeFromExponent(aMessage.GetBlockWiseBlockSize());
+                    SuccessOrExit(error = aTransmitHook(aContext, buf, aMessage.GetBlockWiseBlockNumber() * bufLen,
+                                                        &bufLen, &moreBlocks));
+                    SuccessOrExit(error = aMessage.Append(buf, bufLen));
+
+                    // Block-Wise messages always have to be confirmable
+                    if (aMessage.IsNonConfirmable())
+                    {
+                        aMessage.SetType(OT_COAP_TYPE_CONFIRMABLE);
+                    }
+                }
+            }
+        }
+
+        aMessage.SetMessageId(mMessageId++);
+        break;
+    }
+
+    aMessage.Finish();
+
+    if (aMessage.IsConfirmable())
+    {
+        // Create a copy of entire message and enqueue it.
+        copyLength = aMessage.GetLength();
+    }
+    else if (aMessage.IsNonConfirmable() && (aHandler != NULL))
+    {
+        // As we do not retransmit non confirmable messages, create a
+        // copy of header only, for token information.
+        copyLength = aMessage.GetOptionStart();
+    }
+
+    if (copyLength > 0)
+    {
+        Metadata metadata;
+
+#if OPENTHREAD_CONFIG_COAP_OBSERVE_API_ENABLE
```

Comment:
Can we avoid duplicating this code? Is it possible to integrate the blockwise code into the existing `SendMessage()` method?